tests/test_io.py

```diff
@@ -4,29 +4,141 @@
-        config = {}
 
-        stream = Stream(format=format, start_at=start_at, config=config)
+def content_mock(message_model):
+    """Mock content to pass during the message_model creation since MagicMock()
+    is unable to mock __init__ of the model dataclass in tests.
+    """
+    content = {}
+    for field in fields(message_model):
+        content.update({field.name: "test"})
+    return content
 
-        # verify defaults are stored correctly
-        assert stream._options["format"] == format
-        assert stream._options["start_at"] == start_at
-        assert stream._options["config"] == config
+# test the subscribe classifier for each message format
 
-        with stream.open(broker_url, "w") as s:
+
+@pytest.mark.parametrize("message", [
+    {"format": "voevent", "content": content_mock(VOEvent)},
+    {"format": "circular", "content": content_mock(GCNCircular)},
+    {"format": "blob", "content": content_mock(MessageBlob)},
+    {"format": "other", "content": "other"},
+    ["wrong_datatype"],
+    {"wrong_key": "value"},
+])
+def test_deserialize(message, message_parameters_dict):
+
+    # test a non-dict message
+    if not isinstance(message, dict):
+        with pytest.raises(ValueError):
+            test_model = io.Deserializer.deserialize(message)
+        return
+    # test a dict message with wrong key values
+    elif not (("format" in message) and ("content" in message)):
+        with pytest.raises(ValueError):
+            test_model = io.Deserializer.deserialize(message)
+        return
+
+    message_format = message["format"]
+    message_content = message["content"]
+
+    # test an invalid format
+    if message_format not in message_parameters_dict:
+        with pytest.raises(ValueError):
+            test_model = io.Deserializer.deserialize(message)
+        return
+
+    # load parameters from conftest for valid formats
+    message_parameters = message_parameters_dict[message_format]
+    model_name = message_parameters["model_name"]
+    expected_model = message_parameters["expected_model"]
+
+    # test valid formats
+    with patch(f"hop.models.{model_name}", MagicMock()):
+        test_model = io.Deserializer.deserialize(message)
+
+    # verify the message is classified properly
+    assert isinstance(test_model, expected_model)
+
+
+def test_stream_read(circular_msg, circular_text):
+    with patch("hop.io._Consumer", MagicMock()) as mock_consumer:
+        mock_consumer.stream.return_value = [{'hey', 'you'}]
+
+        broker_url1 = "kafka://hostname:port/gcn"
+        broker_url2 = "kafka://hey@hostname:port/gcn"
+        start_at = io.StartPosition.EARLIEST
+        persist = False
+
+        stream = io.Stream(persist=persist)
+
+        with stream.open(broker_url1, "r", start_at=start_at) as s:
+            for msg in s:
+                continue
+
+        with stream.open(broker_url2, "r", start_at=start_at) as s:
+            for msg in s:
+                continue
```

Comment:
What does this test actually test? I think all that's _really_ getting tested is that `stream.open` does not raise, and that it returns the contents of `mock_consumer.stream` (in other words, that you're mocking works. One issue with using mocks is that you can wind up testing that your mocking library is working, but not whether your code is correct.

Maybe that's sufficient here, but it's a little fishy to me - it doesn't test that `start_at` or `persist` or the URL are handled correctly, just that these options don't raise an exception.

Some ways you could improve this situation:
 - You could write a complicated mock. But I think you'll just be testing your mock code, as I said.
 - You could write integration tests which actually start up a Kafka broker in a container. But these tests can take a long time to execute and there's a lot of complexity in setting them up.
 - You could refactor code until it's unit-testable without mocks. For example, you might break up the `open` function in `io.py` to an `_open_reader` and `_open_writer`; your test of `open` is just whether it calls the right function based on the `mode` argument, and then you thoroughly test each of those new functions. But it's not clear whether this is actually an improvement to the code, other than making it more unit-testable. Obsessive dedication to testability isn't always right!
 - You could just keep the code as-is, relatively lightly tested, relying on your `test_cli` high-level tests rather than expecting unit tests to do much for you. This is probably what I would do.

Anyway, that's a long comment just to end up at "nothing to do here" but I think it's an interesting design question, and you might want to take one of those other paths.

Reply:
> What does this test actually test? I think all that's really getting tested is that stream.open does not raise, and that it returns the contents of mock_consumer.stream (in other words, that your mocking works). One issue with using mocks is that you can wind up testing that your mocking library is working, but not whether your code is correct.

> Maybe that's sufficient here, but it's a little fishy to me - it doesn't test that start_at or persist or the URL are handled correctly, just that these options don't raise an exception.

Yeah, that's right. All this test does is check that `stream.open` doesn't raise. It's better than nothing, but far from satisfactory. It could be improved by checking that certain calls were made to `_Consumer.stream` like you mentioned, but still doesn't feel sufficient. The biggest issue that I've encountered here is that because `_Consumer` subclasses `adc.Consumer` and extends the behavior of `stream()`, I'm not able to mock the parent's class' `stream()`. Maybe not at all, maybe not without considerable work. That would be better because then I can mock sufficiently upstream that I could test the right behavior sufficiently assuming that anything living in `adc` is correct. I kinda went down the refactoring road a bit to split up all the work into these pack/unpack static methods that I could test in isolation, but didn't go any further than that.

I guess the only two things this actually tests is if there's a bug introduced that egregious enough that it raises or the API has changed enough where it would start to fail. Beyond that...nah.

Out of the improvements you suggested, I think a combination of adding integration tests and letting the CLI tests handle more of the heavy lifting sound the most reasonable to me. Adding in integration tests is probably out of scope for this PR but would be good to raise an issue for and tackle soon. Does that sound reasonable?

Reply:
Yep, I think that sounds right to me too :+1: .